File: backend-java/src/main/java/mx/codemx/modules/evaluacion/EvaluacionLocalStrategy.java
package mx.codemx.modules.evaluacion;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.TimeUnit;
import mx.codemx.modules.retos.CasoPrueba;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component
public class EvaluacionLocalStrategy implements EvaluacionStrategy {

    private static class TiempoAgotadoException extends RuntimeException {
        TiempoAgotadoException(String mensaje) {
            super(mensaje);
        }
    }

    private final int timeoutMs;

    public EvaluacionLocalStrategy(@Value("${evaluador.timeout-ms:5000}") int timeoutMs) {
        this.timeoutMs = timeoutMs;
    }

    @Override
    public ResultadoEvaluacion ejecutar(SolicitudEvaluacion solicitud, List<CasoPrueba> casos) {
        if (casos.isEmpty()) {
            return new ResultadoEvaluacion("ERROR_EN_EJECUCION",
                    "El reto no tiene casos de prueba configurados", 0);
        }

        long inicio = System.currentTimeMillis();

        for (CasoPrueba caso : casos) {
            String entrada = caso.getInputData() == null ? "" : caso.getInputData();
            try {
                String salida = ejecutarPython(solicitud.codigoFuente(), entrada);
                String esperado = caso.getOutputEsperado() == null ? "" : caso.getOutputEsperado().trim();

                if (!salida.equals(esperado)) {
                    return new ResultadoEvaluacion("INCORRECTO",
                            "Salida incorrecta para la entrada: " + entrada.replace("\n", " | "),
                            System.currentTimeMillis() - inicio);
                }
            } catch (TiempoAgotadoException e) {
                return new ResultadoEvaluacion("TIEMPO_LIMITE_EXCEDIDO", null, timeoutMs);
            } catch (Exception e) {
                return new ResultadoEvaluacion("ERROR_EN_EJECUCION", e.getMessage(),
                        System.currentTimeMillis() - inicio);
            }
        }

        return new ResultadoEvaluacion("ACEPTADO", null, System.currentTimeMillis() - inicio);
    }

    private String ejecutarPython(String codigo, String stdin) throws IOException, InterruptedException {
        Path archivo = Files.createTempFile("codemx-", ".py");
        try {
            Files.writeString(archivo, codigo, StandardCharsets.UTF_8);

            Process proc = new ProcessBuilder("python3", archivo.toString()).start();

            if (!stdin.isBlank()) {
                try (OutputStream in = proc.getOutputStream()) {
                    in.write(stdin.getBytes(StandardCharsets.UTF_8));
                }
            } else {
                proc.getOutputStream().close();
            }

            var stdoutFuture = leerAsync(proc.getInputStream());
            var stderrFuture = leerAsync(proc.getErrorStream());

            if (!proc.waitFor(timeoutMs, TimeUnit.MILLISECONDS)) {
                proc.descendants().forEach(ProcessHandle::destroyForcibly);
                proc.destroyForcibly();
                throw new TiempoAgotadoException("TLE");
            }

            String stdout = stdoutFuture.join();
            String stderr = stderrFuture.join();

            if (proc.exitValue() != 0) {
                String primeraLinea = Arrays.stream(stderr.split("\n"))
                        .filter(l -> !l.isBlank())
                        .filter(l -> !l.startsWith("  "))
                        .reduce((primera, ultima) -> ultima)
                        .map(String::trim)
                        .orElse("Error en tiempo de ejecución");
                throw new IOException(primeraLinea);
            }

            return stdout.trim();
        } finally {
            Files.deleteIfExists(archivo);
        }
    }

    private static java.util.concurrent.CompletableFuture<String> leerAsync(InputStream stream) {
        return java.util.concurrent.CompletableFuture.supplyAsync(() -> {
            try (stream) {
                return new String(stream.readAllBytes(), StandardCharsets.UTF_8);
            } catch (IOException e) {
                return "";
            }
        });
    }
}

File: backend/Modules/Evaluacion/EvaluacionLocalStrategy.cs
using System.Diagnostics;
using System.Text;
using CodeMX.Api.Modules.Retos;

namespace CodeMX.Api.Modules.Evaluacion;

/// <summary>
/// Estrategia concreta (Strategy): ejecuta el código localmente con un proceso python3,
/// respetando el mismo timeout. Sirve de respaldo mientras el Servicio Python externo
/// no esté disponible.
/// </summary>
public class EvaluacionLocalStrategy : IEvaluacionStrategy
{
    private readonly int _timeoutMs;

    public EvaluacionLocalStrategy(IConfiguration config)
    {
        _timeoutMs = int.TryParse(config["Evaluador:TimeoutMs"], out var t) ? t : 5000;
    }

    public Task<ResultadoEvaluacion> EjecutarAsync(SolicitudEvaluacion solicitud, List<CasoPrueba> casos)
        => Task.FromResult(Ejecutar(solicitud.CodigoFuente, casos));

    private ResultadoEvaluacion Ejecutar(string codigo, List<CasoPrueba> casos)
    {
        if (casos.Count == 0)
            return new ResultadoEvaluacion("ERROR_EN_EJECUCION", "El reto no tiene casos de prueba configurados", 0);

        var inicio = DateTime.UtcNow;

        foreach (var caso in casos)
        {
            try
            {
                var salida = EjecutarPython(codigo, caso.InputData ?? "");
                var esperado = (caso.OutputEsperado ?? "").Trim();

                if (salida != esperado)
                    return new ResultadoEvaluacion("INCORRECTO",
                        "Salida incorrecta para la entrada: " + (caso.InputData ?? "").Replace("\n", " | "),
                        (long)(DateTime.UtcNow - inicio).TotalMilliseconds);
            }
            catch (TimeoutException)
            {
                return new ResultadoEvaluacion("TIEMPO_LIMITE_EXCEDIDO", null, _timeoutMs);
            }
            catch (Exception e)
            {
                return new ResultadoEvaluacion("ERROR_EN_EJECUCION", e.Message,
                    (long)(DateTime.UtcNow - inicio).TotalMilliseconds);
            }
        }

        return new ResultadoEvaluacion("ACEPTADO", null, (long)(DateTime.UtcNow - inicio).TotalMilliseconds);
    }

    private string EjecutarPython(string codigo, string stdin)
    {
        var tempFile = Path.GetTempFileName();
        var pyFile = Path.ChangeExtension(tempFile, ".py");
        File.Move(tempFile, pyFile);
        try
        {
            File.WriteAllText(pyFile, codigo);

            var psi = new ProcessStartInfo
            {
                FileName = "python3",
                Arguments = pyFile,
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false
            };

            using var proc = Process.Start(psi)!;

            if (!string.IsNullOrWhiteSpace(stdin))
            {
                proc.StandardInput.Write(stdin);
                proc.StandardInput.Close();
            }

            var stdoutTask = proc.StandardOutput.ReadToEndAsync();
            var stderrTask = proc.StandardError.ReadToEndAsync();

            if (!proc.WaitForExit(_timeoutMs))
            {
                proc.Kill(entireProcessTree: true);
                throw new TimeoutException("TLE");
            }

            var stdout = stdoutTask.GetAwaiter().GetResult();
            var stderr = stderrTask.GetAwaiter().GetResult();

            if (proc.ExitCode != 0)
            {
                var primeraLinea = stderr
                    .Split('\n', StringSplitOptions.RemoveEmptyEntries)
                    .LastOrDefault(l => !l.StartsWith("  "))?.Trim()
                    ?? "Error en tiempo de ejecución";
                throw new Exception(primeraLinea);
            }

            return stdout.Trim();
        }
        finally
        {
            if (File.Exists(pyFile)) File.Delete(pyFile);
        }
    }
}
